mws/utils/params.py

```diff
@@ -0,0 +1,231 @@
+def enumerate_keyed_param(param, values):
+    """Given a param string and a dict of values, returns a flat dict of keyed, enumerated params.
+    Each dict in the values list must pertain to a single item and its data points.
+
+    Example:
+        param = "InboundShipmentPlanRequestItems.member"
+        values = [
+            {'SellerSKU': 'Football2415',
+            'Quantity': 3},
+            {'SellerSKU': 'TeeballBall3251',
+            'Quantity': 5},
+            ...
+        ]
+
+    Returns:
+        {
+            'InboundShipmentPlanRequestItems.member.1.SellerSKU': 'Football2415',
+            'InboundShipmentPlanRequestItems.member.1.Quantity': 3,
+            'InboundShipmentPlanRequestItems.member.2.SellerSKU': 'TeeballBall3251',
+            'InboundShipmentPlanRequestItems.member.2.Quantity': 5,
+            ...
+        }
+    """
+    if not isinstance(values, (list, tuple, set)):
+        # If it's a single value, convert it to a list first
+        values = [values]
+    if not any(values):
+        # Shortcut for empty values
+        return {}
+    if not param.endswith("."):
+        # Ensure the enumerated param ends in '.'
+        param += "."
+    for val in values:
+        # Every value in the list must be a dict.
+        if not isinstance(val, dict):
+            # Value is not a dict: can't work on it here.
+            raise ValueError(
+                (
+                    "Non-dict value detected. "
+                    "`values` must be a list, tuple, or set; containing only dicts."
+                )
+            )
+    params = {}
+    for idx, val_dict in enumerate(values):
+        # Build the final output.
+        params.update(
+            {
+                "{param}{idx}.{key}".format(param=param, idx=idx + 1, key=k): v
+                for k, v in val_dict.items()
+            }
+        )
+    return params
+
+
+def dict_keyed_param(param, dict_from):
+    """Given a param string and a dict, returns a flat dict of keyed params without enumerate.
+
+    Example:
+        param = "ShipmentRequestDetails.PackageDimensions"
+        dict_from = {'Length': 5, 'Width': 5, 'Height': 5, 'Unit': 'inches'}
+
+    Returns:
+        {
+            'ShipmentRequestDetails.PackageDimensions.Length': 5,
+            'ShipmentRequestDetails.PackageDimensions.Width': 5,
+            'ShipmentRequestDetails.PackageDimensions.Height': 5,
+            'ShipmentRequestDetails.PackageDimensions.Unit': 'inches',
+            ...
+        }
+    """
+    params = {}
+
+    if not param.endswith("."):
+        # Ensure the enumerated param ends in '.'
+        param += "."
+    for k, v in dict_from.items():
+        params.update({"{param}{key}".format(param=param, key=k): v})
+    return params
+
+
+def clean_params_dict(params):
+    """Clean multiple param values in a dict, returning a new dict
+    containing the original keys and cleaned values.
+    """
+    # silently remove parameter where values are empty
```

Comment:
could this actually be dangerous removing empty values silently.

Would it be better to emit a log message here (probably a future consideration) that the developer will see emitted so they don't end up chasing through the code to work out why their input and output differ?

Reply:
Opted to adjust this by separating duties of that function. See new utility, `remove_empty_param_keys`, which takes over that task. That should make it more obvious what the code is doing, for those who want to trace the source.